src/main/cpp/RLBotInterface/BallPrediction/inc/mat.h

```diff
@@ -0,0 +1,350 @@
+    }
+  }
+  return sqrt(sum);
+}
+
+template < int n >
+inline float tr(const mat < n, n > & A) {
+  float sum = 0.0;
+  for (int i = 0; i < n; i++) {
+    sum += A(i,i);
+  }
+  return sum;
+}
+
+inline float det(const mat < 2, 2 > & A){
+  return A(0,0)*A(1,1)-A(0,1)*A(1,0);
+}
+
+inline float det(const mat < 3, 3 > & A){
+  return A(0,0)*A(1,1)*A(2,2)+A(0,1)*A(1,2)*A(2,0)+
+         A(0,2)*A(1,0)*A(2,1)-A(0,0)*A(1,2)*A(2,1)-
+         A(0,1)*A(1,0)*A(2,2)-A(0,2)*A(1,1)*A(2,0);
+}
+
+inline float det(const mat < 4, 4 > & A){
+  return A(0,3)*A(1,2)*A(2,1)*A(3,0)-A(0,2)*A(1,3)*A(2,1)*A(3,0)-
+         A(0,3)*A(1,1)*A(2,2)*A(3,0)+A(0,1)*A(1,3)*A(2,2)*A(3,0)+
+         A(0,2)*A(1,1)*A(2,3)*A(3,0)-A(0,1)*A(1,2)*A(2,3)*A(3,0)-
+         A(0,3)*A(1,2)*A(2,0)*A(3,1)+A(0,2)*A(1,3)*A(2,0)*A(3,1)+
+         A(0,3)*A(1,0)*A(2,2)*A(3,1)-A(0,0)*A(1,3)*A(2,2)*A(3,1)-
+         A(0,2)*A(1,0)*A(2,3)*A(3,1)+A(0,0)*A(1,2)*A(2,3)*A(3,1)+
+         A(0,3)*A(1,1)*A(2,0)*A(3,2)-A(0,1)*A(1,3)*A(2,0)*A(3,2)-
+         A(0,3)*A(1,0)*A(2,1)*A(3,2)+A(0,0)*A(1,3)*A(2,1)*A(3,2)+
+         A(0,1)*A(1,0)*A(2,3)*A(3,2)-A(0,0)*A(1,1)*A(2,3)*A(3,2)-
+         A(0,2)*A(1,1)*A(2,0)*A(3,3)+A(0,1)*A(1,2)*A(2,0)*A(3,3)+
+         A(0,2)*A(1,0)*A(2,1)*A(3,3)-A(0,0)*A(1,2)*A(2,1)*A(3,3)-
+         A(0,1)*A(1,0)*A(2,2)*A(3,3)+A(0,0)*A(1,1)*A(2,2)*A(3,3);
+}
+
+template < int n >
+inline float I1(const mat < n, n > & A) {
+  return tr(A);
+}
+
+template < int n >
+inline float I2(const mat < n, n > & A) {
+  float trA = tr(A);
+  float trAA = 0.0;
+  for (int i = 0; i < n; i++) {
+    for (int j = 0; j < n; j++) {
+      trAA += A(i,j) * A(j,i);
+    }
+  }
+  return 0.5 * (trAA - trA * trA);
+}
+
+template < int n >
+inline float I3(const mat < n, n > & A) {
+  return det(A);
+}
+
+template < int n >
+inline mat < n, n > dev(const mat < n, n > & A) {
+  auto A_dev = A; 
+  float trA = tr(A);
+  for (int i = 0; i < n; i++) {
+    A_dev(i,i) -= (trA / n);
+  } 
+  return A_dev;
+}
+
+inline mat < 2, 2 > inv(const mat < 2, 2 > & A){
+
+  mat < 2, 2 > invA;
+
+  float inv_detA = 1.0f / det(A);
+  
+  invA(0,0) =  A(1,1) * inv_detA;
+  invA(0,1) = -A(0,1) * inv_detA;
+  invA(1,0) = -A(1,0) * inv_detA;
+  invA(1,1) =  A(0,0) * inv_detA;
+
+  return invA;
+
+}
+
+inline mat < 3, 3 > inv(const mat < 3, 3 > & A){
+
+  mat < 3, 3 > invA;
+
+  float inv_detA = 1.0f / det(A);
+  
+  invA(0,0) = (A(1,1)*A(2,2) - A(1,2)*A(2,1)) * inv_detA;
+  invA(0,1) = (A(0,2)*A(2,1) - A(0,1)*A(2,2)) * inv_detA;
+  invA(0,2) = (A(0,1)*A(1,2) - A(0,2)*A(1,1)) * inv_detA;
+  invA(1,0) = (A(1,2)*A(2,0) - A(1,0)*A(2,2)) * inv_detA;
+  invA(1,1) = (A(0,0)*A(2,2) - A(0,2)*A(2,0)) * inv_detA;
+  invA(1,2) = (A(0,2)*A(1,0) - A(0,0)*A(1,2)) * inv_detA;
+  invA(2,0) = (A(1,0)*A(2,1) - A(1,1)*A(2,0)) * inv_detA;
+  invA(2,1) = (A(0,1)*A(2,0) - A(0,0)*A(2,1)) * inv_detA;
+  invA(2,2) = (A(0,0)*A(1,1) - A(0,1)*A(1,0)) * inv_detA;
+
+  return invA;
+
+} 
+
+inline mat < 4, 4 > inverse(const mat < 4, 4 > & A){
```

Comment:
also can we use the word matrix instead of mat?

Reply:
many of these naming conventions are stolen wholesale from https://glm.g-truc.net/0.9.4/api/index.html